contracts/Governance.sol

```diff
@@ -0,0 +1,256 @@
+        proposals[_id].uriDescription = _uriDescription;
+        proposals[_id].uriArguments = _uriArguments;
+
+        proposals[_id].descriptionHash = keccak256(proposals[_id].uriDescription);
+        proposals[_id].argumentsHash = keccak256(proposals[_id].uriArguments);
+
+        emit ProposalCreated(_id, _destination);
+
+        requestRegisteringProposal(_id);
+    }
+
+
+    /** @dev Request registering a proposal to the proposal list
+     *  @param _id ID of a proposal
+     */
+    function requestRegisteringProposal(bytes32 _id) public payable onlyWhenProposalIsNew(_id) {
+        proposalList.requestRegistration.value(msg.value)(_id);
+        emit ProposalRequestedToRegister(_id);
+    }
+
+
+    /** @dev Put proposal to support voting only when a new proposal is permitted.
+     *  @param _id ID of a proposal
+     */
+    function putProposalToSupport(bytes32 _id) public onlyWhenProposalIsNew(_id) {
+        require(proposalList.isPermitted(_id), "Only when proposal is permitted.");
+
+        Proposal storage proposal = proposals[_id];
+
+        address cloneToken = pinakion.createCloneToken({_cloneTokenName: quorumTokenName, _cloneDecimalUnits: DECIMALS, _cloneTokenSymbol: quorumTokenSymbol, _snapshotBlock: block.number, _transfersEnabled: true});
+        proposal.quorumToken = MiniMeToken(cloneToken);
+        proposal.quorumToken.changeController(tokenController);
+
+        proposal.state = ProposalState.PutToSupport;
+
+        emit ProposalPutToSupport(_id);
+    }
+
+
+    /** @dev Calculate and return required quorum for a given proposal.
+     *  @param _id ID of a proposal
+     */
+    function getRequiredQuorum(bytes32 _id) public view onlyWhenProposalPutToSupport(_id) returns (uint effectiveQuorum){
+        uint numberOfDividePeriodsPassed = (block.timestamp - lastTimeQuorumReached) / quorumDivideTime;
+        effectiveQuorum = currentProposalQuorum * proposals[_id].quorumToken.totalSupply() / (2 ** numberOfDividePeriodsPassed) / 100;
+    }
+
+
+    /** @dev Put given proposal to vote.
+     *  @param _id ID of a proposal
+     */
+    function putProposalToVote(bytes32 _id) public onlyWhenProposalPutToSupport(_id) {
+        require(proposals[_id].quorumToken.balanceOf(supportDeposit) >= getRequiredQuorum(_id), "Only when propsal has quorum.");
+
+        Proposal storage proposal = proposals[_id];
+
+        proposal.whenPutToVote = block.timestamp;
+
+        address cloneToken = pinakion.createCloneToken({_cloneTokenName: voteTokenName, _cloneDecimalUnits: DECIMALS, _cloneTokenSymbol: voteTokenSymbol, _snapshotBlock: block.number, _transfersEnabled: true});
+        proposal.voteToken = MiniMeToken(cloneToken);
+        proposal.voteToken.changeController(tokenController);
+
+        proposal.state = ProposalState.PutToVote;
+
+        emit ProposalPutToVote(_id);
+
+        lastTimeQuorumReached = block.timestamp; // Necessary when calculating required quorum as it is halved periodically.
```

Comment:
What happens when more than one proposal is being voted on at the same time? Does resetting these global variables effect the first vote?

Reply:
This is about quorum (support) voting. When a proposal reaches the quorum, the timer resets.

Quorum requirement has a half-life, if in x seconds no proposal reaches quorum, requirement gets halved.

- Let's say there are two proposals who have %35 support at the moment. 
- Base quorum requirement is %60 but it got halved one time, so it's %30 now. 

In this case there is a race, first call to `putProposalToVote()` wins and resets quorum requirement to %60.

Reply:
Hmm, I understand what you mean, yes the first one will change the global variables and affect the second one.

Should we keep it constant after proposal creation? If so I can keep those variables in the struct and initialize them when created.

Reply:
I think it's OK that it affects the second one. It should reflect the latest activity levels of members.

Reply:
Actually, from the spec:
> The quorum for this proposal is saved such that quorum modifications do not affect it.

Reply:
> Actually, from the spec:
> 
> > The quorum for this proposal is saved such that quorum modifications do not affect it.

So we should keep it constant after proposal creation, OK.